examples/interpreter-demo/serialize.js

```diff
@@ -0,0 +1,411 @@
+/**
+ * @license
+ * Copyright 2017 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Saving and restoring the state of a JS-Intepreter.
+ * @author fraser@google.com (Neil Fraser)
+ */
+'use strict';
+
+
+// Constructors for objects within Acorn.
+var NODE_CONSTRUCTOR;
+var NODE_LOC_CONSTRUCTOR;
+var LINE_LOC_CONSTRUCTOR;
+
+function recordAcornConstructons(interpreter) {
+  // Constructors for objects within Acorn.
+  NODE_CONSTRUCTOR = interpreter.ast.constructor;
+  NODE_LOC_CONSTRUCTOR = interpreter.ast.loc &&
+      interpreter.ast.loc.constructor;
+  LINE_LOC_CONSTRUCTOR = interpreter.ast.loc &&
+      interpreter.ast.loc.end.constructor;
+}
+
+function deserialize(json, interpreter) {
+  function decodeValue(value) {
+    if (value && typeof value === 'object') {
+      var data;
+      if ((data = value['#'])) {
+       // Object reference: {'#': 42}
+       value = objectList[data];
+        if (!value) {
+          throw ReferenceError('Object reference not found: ' + data);
+        }
+        return value;
+      }
+      if ((data = value['Number'])) {
+        // Special number: {'Number': 'Infinity'}
+        return Number(data);
+      }
+      if ((data = value['Value'])) {
+        // Special value: {'Value': 'undefined'}
+        if (value['Value'] === 'undefined') {
+          return undefined;
+        }
+      }
+    }
+    return value;
+  }
+  var stack = interpreter.getStateStack();
+  if (!Array.isArray(json)) {
+    throw TypeError('Top-level JSON is not a list.');
+  }
+  if (!stack.length) {
+    // Require native functions to be present.
+    throw Error('Interpreter must be initialized prior to deserialization.');
+  }
+  recordAcornConstructons(interpreter);
+  var LOC_REGEX = /^(\d*):(\d*)-(\d*):(\d*) ?(.*)$/;
+  // Find all native functions in existing interpreter.
+  var objectList = [];
+  objectHunt_(stack, objectList);
+  var functionHash = Object.create(null);
+  for (var i = 0; i < objectList.length; i++) {
+    if (typeof objectList[i] === 'function') {
+      functionHash[objectList[i].id] = objectList[i];
+    }
+  }
+  // First pass: Create object stubs for every object.
+  objectList = [];
+  for (var i = 0; i < json.length; i++) {
```

Comment:
Break these passes out into separate functions for readability.

Reply:
Done.